# Revision History

# 4.0.1-SNAPSHOT

*This was last updated on 11/11/2020*

* Migration to Java 11. **You do need Java >= 11 to make knetminer working** If you use Docker, our images
  use the right Java version automatically.

## Internals
* Significant review of the codebase file structure and names
* Third-party dependency updates
* Big refactoring of the `OndexServiceProvider` component
* Docker-base removed, now there are only two image levels


# 4.0

## What's new?

1. New KnetMiner branding
    * KnetMiner 4.0 sees a major revamp with an all-new [website](https://knetminer.com/wordpress/) and new branding, logo, fonts, and separated sub- products and services with their own logo, content pages and demos.
    * KnetMiner now on [wikipedia](https://en.wikipedia.org/wiki/KnetMiner).

2. New COVID-19 KnetMiner
    * We have developed a new [COVID-19 KnetMiner](www.knetminer.org/COVID-19). Read the full press release [here](https://www.rothamsted.ac.uk/news/rothamsted-answers-white-house-call-coronavirus-data-help) and check out COVID-19 KnetMiner's [feature appearance in Nature](https://www.nature.com/articles/d41586-020-01733-7).

3. Save your knetworks in the all-new KnetSpace
    * KnetMiner users can now save their knowledge networks on [KnetSpace (beta)](https://knetminer.com/beta/knetspace/),  our new collaborative knowledge discovery platform to manage KnetMiner knetworks.
    * New login feature added to KnetMiner with a “Sign in '' button in the header. Users can now sign-in using their KnetSpace username and password (sign-up is very quick). In future, we will add Google login and similar.
    * Previously, users could perform a keyword or gene list search in KnetMiner, generate an initial network for some genes and export a PNG image. As of KnetMiner 4.0, you can save it to your KnetSpace and can edit it anytime in future and share it with your students, colleagues or collaborators to help you find interesting links or curate the network collaboratively (eg hide certain nodes/edges, add labels, change layout). 
   * You can give people permissions to either edit your original network or clone it to their own KnetSpace. Once you are happy with a network you can tweet the link or share the interactive network on your own website. 
   * To save networks, simply click the new “Save network” button on the KnetMiner Network View tab and provide a name and description. This will quickly save the visible network, its layout and metadata information such as owner, date, version, source, genes, keywords used, etc. to your KnetSpace account.
![knetspace_git](https://user-images.githubusercontent.com/47788638/84669863-4254fd80-af1d-11ea-8b07-9c8ad349fc61.gif)

4. Free and Pro user subscription plans
    * We have added Free and Pro subscription plans. Sign-up to KnetMiner and KnetSpace is for free. The Free plan provides all standard KnetMiner features, plus allows users to store a limited number of user knetworks to KnetSpace. The Pro plan provides greater storage and allows users to make their saved Knetworks private. Choose a plan that’s right for you.

5. Neo4j-Cypher integration (beta)
   * The graph traversal code was re-engineered to support the Cypher graph query language. This is the first major milestone to transition towards an alternative graph database backend. Cypher is a more expressive and more standardised query language. This feature is still experimental and requires a separate Neo4j database installation, based on our RDF and Neo4j export tools. The default KnetMiner 4.0 option is to use the old semantic motif search.

References and further details:
   * overview of the new Knetminer architecture, https://www.degruyter.com/view/journals/jib/15/3/article-20180023.xml
   * Details on the Knetminer backend and sample Cypher motif queries, https://www.slideshare.net/mbrandizi/knetminer-backend-training-nov-2018-124145829
   * Real use cases can be found on our dataset files, [example](https://github.com/Rothamsted/knetminer/blob/master/datasets/wheat-beta/ws/cypher-queries.txt). More documentation will be released soon.

## Bug fixes and software improvements

### Search
 * Searches pertaining to a gene identifier, using the Gene List, are now *exact*.

### API
Several new APIs have been implemented now, including the following:
   * API for KnetSpace host url address, as defined in the POM by variable knetminer.knetSpaceHost with the API URL given as /ws/{$speciesName}/ksHost 
   * API for geneCount added which provides the size of the subgraph (number of nodes, edges) generated by KnetMiner for a given gene list. Given as /ws/{$speciesName}/geneCount?list=
   * DataSource information of the knowledge graph including source organisation, database version,  provider, species tax ID, date of creation, and species name, is accessible via the /ws/{$speciesName}/dataSource API. 

### KnetMiner back-end
   * KnetMiner can now be customised and deployed with a subset of genes. For example, we limited the COVID-19 KnetMiner to call 8,000 human genes as the seed. The path to the seed gene list can be defined in the main KnetMiner POM file variable knetminer.backend.seedGenesFile. This means only the seed genes will appear in the KnetMiner search index. The default is to index all genes. Refer to this [commit](https://github.com/Rothamsted/knetminer/commit/fe16a32196cdb18025ce11728d02aa03799f7c25).
   * Users may now define the number of Publications exported in knowledge subgraphs by altering the main POM configuration variable defaultExportedPublicationCount. The default is set to 20 most recent publications. 
   * The unconnected filter prior to JSON export was extended with additional concept classes 
   * Metadata information for the knowledge graph can now be defined in the KnetMiner dataset/species POM.

### Network view
  * Updated to [KnetMaps.js](https://knetminer.rothamsted.ac.uk/KnetMaps/) v2.0.2. 
  * Save PNG and JSON have been replaced with a ‘Save Knetwork’ button, as described in the KnetSpace integration section.
   * More data sources URLs added to support rendering of new data types.
   * Custom branding and fonts packaged for KnetMaps as well via gulp-configs.
   * Keyword highlighting logic updated and limited only to primary colours and some shades.

   
   
# 3.2

## New features/enhancements
* New genepage design using KnetMiner look & feel
* New analytics logger for genome, network and genepage modes
* Network View legend was enhanced
* Release Notes stats table was expanded with further graph analytics 
* New search button
* Updated footer